pimd/pim6_mld.c

```diff
@@ -537,6 +537,8 @@ static void gm_packet_drop(struct gm_packet_state *pkt, bool trace)
 		if (!sg)
 			continue;
 
+		THREAD_OFF(sg->t_sg_expire);
```

Comment:
Why doesn't this belong in gm_packet_sg_drop?  Or in gm_packet_free()?  We call gm_packet_sg_drop all over the place and delete the packet.  Isn't this just once place that the data structure could be freed?  I'll defer to @eqvinox though if he disagrees here

Reply:
It actually belongs in `gm_sg_update`. The SG expiry timer is, well, for the SG. Multiple gm_packet may be pointing to the same SG, and we need to cancel it only if the last one went away.

So, yeah, this fix is not correct, it can't be cancelled unconditionally, but your suggestion is in the wrong direction too :laughing:. I need to look at the state machine to check which specific path in `gm_sg_update` this should hit, if I'm not mistaken the fix is going to be in the `desired == GM_SG_NOINFO` path at the end. `if (has_expired) THREAD_OFF(sg->t_sg_expire);` before the `assertf()`.